<a href="https://colab.research.google.com/github/Archandra12/Term4-Projects/blob/main/Pandas_testINSAID.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

An attempt to use Pandas DataFrame manipulation to answer a few questions. 

Questions are available at https://notepad.pw/l7nfy292

##Let's begin

import libraries

In [3]:
import numpy as np
import pandas as pd

### Dataset 1. VideoGameSales.csv
The dataset can be imported as pandas dataframe using the following link: https://raw.githubusercontent.com/insaid2018/INSAID-Activities/main/Data%20Detective%20Challenge/VideoGameSales.csv

reference:
https://kanoki.org/2019/03/27/pandas-select-rows-by-condition-and-string-operations/
https://stackoverflow.com/questions/41689722/how-to-select-rows-that-do-not-start-with-some-str-in-pandas/41689928

In [4]:
df_videogames = pd.read_csv('https://raw.githubusercontent.com/insaid2018/INSAID-Activities/main/Data%20Detective%20Challenge/VideoGameSales.csv')

In [ ]:
df_videogames.head(5)

Rank                      Name Platform    Year         Genre Publisher  NA_Sales_Volume  EU_Sales_Volume  JP_Sales_Volume  Other_Sales_Volume
0     1                Wii Sports      Wii  2006.0        Sports  Nintendo            41.49            29.02             3.77                8.46
1     2         Super Mario Bros.      NES  1985.0      Platform  Nintendo            29.08             3.58             6.81                0.77
2     3            Mario Kart Wii      Wii  2008.0        Racing  Nintendo            15.85            12.88             3.79                3.31
3     4         Wii Sports Resort      Wii  2009.0        Sports  Nintendo            15.75            11.01             3.28                2.96
4     5  Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo            11.27             8.89            10.22                1.00

#### 1. Which of the following genres are in the top 5 in JP and NA region based on the total sales volume?

In [6]:
  print(df_videogames.groupby('Genre')['NA_Sales_Volume'].sum().nlargest(7))
  df_videogames.groupby('Genre')['JP_Sales_Volume'].sum().nlargest(7)

Genre
Action          877.83
Sports          683.35
Shooter         582.60
Platform        447.05
Misc            410.24
Racing          359.42
Role-Playing    327.28
Name: NA_Sales_Volume, dtype: float64


Genre
Role-Playing    352.31
Action          159.95
Sports          135.37
Platform        130.77
Misc            107.76
Fighting         87.35
Simulation       63.70
Name: JP_Sales_Volume, dtype: float64

#### 2.Which of the top 5 publishers enjoyed the highest amount of sales volume in the NA region?

In [7]:
 print(df_videogames.groupby('Publisher')['NA_Sales_Volume'].sum().nlargest(5))

Publisher
Nintendo                       816.87
Electronic Arts                595.07
Activision                     429.70
Sony Computer Entertainment    265.22
Ubisoft                        253.43
Name: NA_Sales_Volume, dtype: float64


####4.0 Which of the following platform saw the lowest percentage increase (positive) in global sales volume of games from 2013 to 2014? Note: Global sales can be calculated as the sum of sales of all the regions in the dataset. 



In [9]:
df_videogames['Global_Sales_Volume'] = df_videogames['NA_Sales_Volume'] + df_videogames['EU_Sales_Volume']	+ df_videogames['JP_Sales_Volume']	+ df_videogames['Other_Sales_Volume']

In [10]:
df_videogames_2013 = df_videogames[(df_videogames['Year']== 2013) ].groupby(['Platform', 'Year'])['Global_Sales_Volume'].sum()
df_videogames_2014 = df_videogames[(df_videogames['Year']== 2014) ].groupby(['Platform', 'Year'])['Global_Sales_Volume'].sum()

In [11]:
df_2014 = pd.DataFrame(df_videogames_2014)

In [12]:
df_2013 = pd.DataFrame(df_videogames_2013)

In [13]:
df_2014.reset_index(inplace=True)
df_2013.reset_index(inplace=True)

In [14]:
df_margin = pd.merge(df_2013, df_2014,how='inner',on='Platform')

In [15]:
df_margin['YOY'] = (df_margin['Global_Sales_Volume_y'] - df_margin['Global_Sales_Volume_x'])/df_margin['Global_Sales_Volume_x']

In [16]:
df_margin # 'WiiU'

Platform  Year_x  ...  Global_Sales_Volume_y       YOY
0       3DS  2013.0  ...                  43.12 -0.228070
1        DS  2013.0  ...                   0.02 -0.989796
2        PC  2013.0  ...                  13.33  0.041406
3       PS3  2013.0  ...                  50.97 -0.565917
4       PS4  2013.0  ...                  98.83  2.991519
5       PSP  2013.0  ...                   0.24 -0.924765
6       PSV  2013.0  ...                  14.75  0.167854
7       Wii  2013.0  ...                   4.43 -0.527719
8      WiiU  2013.0  ...                  22.54  0.033471
9      X360  2013.0  ...                  36.43 -0.593234
10     XOne  2013.0  ...                  52.44  1.817840

[11 rows x 6 columns]

#### 13.0 What is the ratio of total sales of Activision in JP to that of EU?

In [17]:
df_videogames[df_videogames['Publisher']=='Activision'].groupby('Publisher')[('JP_Sales_Volume', 'EU_Sales_Volume')].sum()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


JP_Sales_Volume  EU_Sales_Volume
Publisher                                   
Activision             6.54           215.53

In [18]:
activision_JP_EU_ratio = df_videogames[df_videogames['Publisher']=='Activision'].groupby('Publisher')[('JP_Sales_Volume')].sum()/df_videogames[df_videogames['Publisher']=='Activision'].groupby('Publisher')[( 'EU_Sales_Volume')].sum()
activision_JP_EU_ratio

Publisher
Activision    0.030344
dtype: float64

In [ ]:
df_videogames[(df_videogames['Year']== 2013) | (df_videogames['Year']== 2014)].groupby(['Platform', 'Year'])['Global_Sales_Volume'].sum()

Platform  Year  
3DS       2013.0     55.86
          2014.0     43.12
DS        2013.0      1.96
          2014.0      0.02
PC        2013.0     12.80
          2014.0     13.33
PS3       2013.0    117.42
          2014.0     50.97
PS4       2013.0     24.76
          2014.0     98.83
PSP       2013.0      3.19
          2014.0      0.24
PSV       2013.0     12.63
          2014.0     14.75
Wii       2013.0      9.38
          2014.0      4.43
WiiU      2013.0     21.81
          2014.0     22.54
X360      2013.0     89.56
          2014.0     36.43
XOne      2013.0     18.61
          2014.0     52.44
Name: Global_Sales_Volume, dtype: float64

In [ ]:
df_videogames[df_videogames['Year']== 2014]

Rank                                               Name Platform    Year         Genre              Publisher  NA_Sales_Volume  EU_Sales_Volume  JP_Sales_Volume  Other_Sales_Volume  Global_Sales_Volume
44        45                                 Grand Theft Auto V      PS4  2014.0        Action   Take-Two Interactive             3.80             5.81             0.36                2.02                11.99
49        50          Pokemon Omega Ruby/Pokemon Alpha Sapphire      3DS  2014.0  Role-Playing               Nintendo             4.23             3.37             3.08                0.65                11.33
93        94                     Call of Duty: Advanced Warfare      PS4  2014.0       Shooter             Activision             2.80             3.30             0.14                1.37                 7.61
97        98                Super Smash Bros. for Wii U and 3DS      3DS  2014.0      Fighting               Nintendo             3.24             1.35             2.42                0.43                 7.44
108      109                                       Mario Kart 8     WiiU  2014.0        Racing               Nintendo             3.13             2.07             1.27                0.49                 6.96
...      ...                                                ...      ...     ...           ...                    ...              ...              ...              ...                 ...                  ...
16519  16522                         Bakumatsu Rock: Ultra Soul      PSP  2014.0          Misc  Marvelous Interactive             0.00             0.00             0.01                0.00                 0.01
16553  16556                                     Bound By Flame     X360  2014.0  Role-Playing                    NaN             0.00             0.01             0.00                0.00                 0.01
16557  16560                          Dramatical Murder Re:code      PSV  2014.0     Adventure              Nitroplus             0.00             0.00             0.01                0.00                 0.01
16560  16563                                  Hanasaku Manimani      PSV  2014.0     Adventure                    5pb             0.00             0.00             0.01                0.00                 0.01
16567  16570  Fujiko F. Fujio Characters: Great Assembly! Sl...      3DS  2014.0        Action     Namco Bandai Games             0.00             0.00             0.01                0.00                 0.01

[582 rows x 11 columns]

In [ ]:
df_videogames.head()

Rank                      Name Platform    Year         Genre Publisher  NA_Sales_Volume  EU_Sales_Volume  JP_Sales_Volume  Other_Sales_Volume  Global_Sales_Volume
0     1                Wii Sports      Wii  2006.0        Sports  Nintendo            41.49            29.02             3.77                8.46                82.74
1     2         Super Mario Bros.      NES  1985.0      Platform  Nintendo            29.08             3.58             6.81                0.77                40.24
2     3            Mario Kart Wii      Wii  2008.0        Racing  Nintendo            15.85            12.88             3.79                3.31                35.83
3     4         Wii Sports Resort      Wii  2009.0        Sports  Nintendo            15.75            11.01             3.28                2.96                33.00
4     5  Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo            11.27             8.89            10.22                1.00                31.38

#### 21. PlayStation (PS, PS2, PS3, PS4, PSP, and PSV) and XBOX (XB, X360, and XOne) are competitor console platforms from Sony, a Japanese (JP) company, and Microsoft, an American (NA) company respectively. Based on this information, can we check that Playstation (PS, PS2, PS3, PS4, PSP, and PSV) has more game sales in the JP region than XBOX (XB, X360, and XOne)? Likewise, check if XBOX has more game sales in the NA region than Playstation?

In [20]:
PS_sales_JP = df_videogames[df_videogames.Name.str.contains('PS')]['JP_Sales_Volume'].sum()
PS_sales_NA = df_videogames[df_videogames.Name.str.contains('PS')]['NA_Sales_Volume'].sum()
print('PS Comparison: JP sales ',  PS_sales_JP, ' NA Sales', PS_sales_NA)

PS Comparison: JP sales  0.45  NA Sales 5.3


In [21]:
PS_sales_JP = df_videogames[df_videogames.Name.str.contains('X')]['JP_Sales_Volume'].sum()
PS_sales_NA = df_videogames[df_videogames.Name.str.contains('X')]['NA_Sales_Volume'].sum()
print('XBOX Comparison: JP sales ',  PS_sales_JP, ' NA Sales', PS_sales_NA)

XBOX Comparison: JP sales  56.24  NA Sales 93.43


###Dataset 2. market_data.csv
####3. How many customers were attracted from Spain (SP) in at least 5 of the campaigns?

Note: Also, include the "Response" column too as it is also a campaign.

For example, AcceptedCmp1 stands for Campaign 1. Consider the columns that start with "AcceptedCmp". So in all, there are 6 campaigns.

The dataset can be imported as pandas dataframe using the following link: https://raw.githubusercontent.com/insaid2018/INSAID-Activities/main/Data%20Detective%20Challenge/market_data.csv

In [22]:
df_market = pd.read_csv('https://raw.githubusercontent.com/insaid2018/INSAID-Activities/main/Data%20Detective%20Challenge/market_data.csv')
df_market.head()

ID  Year_Birth   Education  ... Response Complain  Country
0   1826        1970  Graduation  ...        1        0       SP
1      1        1961  Graduation  ...        1        0       CA
2  10476        1958  Graduation  ...        0        0       US
3   1386        1967  Graduation  ...        0        0      AUS
4   5371        1989  Graduation  ...        1        0       SP

[5 rows x 28 columns]

In [23]:
df_market[(df_market['Country']=='SP') & (df_market['AcceptedCmp1']==1) & (df_market['AcceptedCmp2']==1) & (df_market['AcceptedCmp3']==1) & (df_market['AcceptedCmp4']==1)]

Empty DataFrame
Columns: [ID, Year_Birth, Education, Marital_Status, Income, Kidhome, Teenhome, Dt_Customer, Recency, MntWines, MntFruits, MntMeatProducts, MntFishProducts, MntSweetProducts, MntGoldProds, NumDealsPurchases, NumWebPurchases, NumCatalogPurchases, NumStorePurchases, NumWebVisitsMonth, AcceptedCmp3, AcceptedCmp4, AcceptedCmp5, AcceptedCmp1, AcceptedCmp2, Response, Complain, Country]
Index: []

#### 4. Which of the following product performed the best in terms of the average amount spent?

For example, MntFishProducts denotes the amount spent on Fish products. Consider only the columns that start with "Mnt". 

In [24]:
dict = {}
for i in df_market.columns:
  if i[:3]=='Mnt':
    print(i, df_market[i].sum())
    dict[i]=df_market[i].sum()
    #print (i)

MntWines 680816
MntFruits 58917
MntMeatProducts 373968
MntFishProducts 84057
MntSweetProducts 60621
MntGoldProds 98609


In [25]:
pd.DataFrame(data=dict, index=[1]).T

1
MntWines          680816
MntFruits          58917
MntMeatProducts   373968
MntFishProducts    84057
MntSweetProducts   60621
MntGoldProds       98609

#### 6. What is the maximum income (in USD) of the married people from IND country having 2 children with Graduation Education born after 1975?

In [26]:
df_market.head()

ID  Year_Birth   Education  ... Response Complain  Country
0   1826        1970  Graduation  ...        1        0       SP
1      1        1961  Graduation  ...        1        0       CA
2  10476        1958  Graduation  ...        0        0       US
3   1386        1967  Graduation  ...        0        0      AUS
4   5371        1989  Graduation  ...        1        0       SP

[5 rows x 28 columns]

In [27]:
df_market[(df_market['Marital_Status']=='Married') & (df_market['Kidhome']==2) & (df_market['Education']=='Graduation')  & (df_market['Country']=='IND')  & (df_market['Year_Birth'] >1975)]['Income'].max()

'$55,357.00 '

#### 8. Which country has the highest number of total purchases?

In [28]:
df_market['total_purchases'] = df_market['NumDealsPurchases'] + df_market['NumWebPurchases'] + df_market['NumCatalogPurchases']	+ df_market['NumStorePurchases']

In [29]:
df_market.groupby('Country')['total_purchases'].sum()

Country
AUS     2314
CA      4101
GER     1788
IND     2099
ME        59
SA      5117
SP     16052
US      1761
Name: total_purchases, dtype: int64

####17. There are rows corresponding to the "Income" column which have missing or null values. Which of the following technique can be used to deal with this issue?

###Just question here we can have our answers based on our expertise!

###19 Which of the following campaign had the least success rate?

For example, AcceptedCmp1 stands for Campaign 1. Consider only the columns that start with "AcceptedCmp". 

In [30]:
df_market.head()

ID  Year_Birth   Education  ... Complain Country  total_purchases
0   1826        1970  Graduation  ...        0      SP               15
1      1        1961  Graduation  ...        0      CA               18
2  10476        1958  Graduation  ...        0      US               11
3   1386        1967  Graduation  ...        0     AUS                4
4   5371        1989  Graduation  ...        0      SP                8

[5 rows x 29 columns]

In [31]:
dict = {}
for i in df_market.columns:
  if i[:-1]=='AcceptedCmp':
    print(i, df_market[i].sum())
    dict[i]=df_market[i].sum()
    #print (i)
dict

AcceptedCmp3 163
AcceptedCmp4 167
AcceptedCmp5 163
AcceptedCmp1 144
AcceptedCmp2 30


{'AcceptedCmp1': 144,
 'AcceptedCmp2': 30,
 'AcceptedCmp3': 163,
 'AcceptedCmp4': 167,
 'AcceptedCmp5': 163}


Assuming appearing of 1 as success rate, Campaign 2 seems least successful.

###Dataset 3. goodreads.csv:
####5. Which of the following parameter (additional) is required to read the goodreads.csv file using pandas?

The dataset can be imported as pandas dataframe using the following link: https://raw.githubusercontent.com/insaid2018/INSAID-Activities/main/Data%20Detective%20Challenge/goodreads.csv

In [32]:
df_goodreads = pd.read_csv('https://raw.githubusercontent.com/insaid2018/INSAID-Activities/main/Data%20Detective%20Challenge/goodreads.csv', )

In [33]:
df_goodreads.describe()

bookID  average_rating  ...  ratings_count  text_reviews_count
count  11123.000000    11123.000000  ...   1.112300e+04        11123.000000
mean   21310.856963        3.934075  ...   1.794285e+04          542.048099
std    13094.727252        0.350485  ...   1.124992e+05         2576.619589
min        1.000000        0.000000  ...   0.000000e+00            0.000000
25%    10277.500000        3.770000  ...   1.040000e+02            9.000000
50%    20287.000000        3.960000  ...   7.450000e+02           47.000000
75%    32104.500000        4.140000  ...   5.000500e+03          238.000000
max    45641.000000        5.000000  ...   4.597666e+06        94265.000000

[8 rows x 6 columns]

####14. How many books can be considered as outliers based on the average rating? Note: Use 1.5 as a threshold value to estimate upper and lower limits using IQR to detect the outliers.

##Missed this one..

####24. How many Spanish authors have written more than three books?

In [37]:
df_count = df_goodreads[df_goodreads['language_code'] == 'spa'].groupby('authors')['bookID'].count().reset_index()

In [35]:
df_count

authors  bookID
0                                  Adolfo Bioy Casares       1
1                          Albert Camus/Manuel de Lope       1
2    Alexander McCall Smith/Marta Torent López de L...       1
3                       Alice Sebold/Aurora Echevarría       1
4                                Amy Tan/Claudia Conde       1
..                                                 ...     ...
159                                    William Goldman       1
160                       Yann Martel/Bianca Southwood       1
161           Z.Z. Packer/María Pérez López de Heredia       1
162                                        Zadie Smith       1
163                                     Åsne Seierstad       1

[164 rows x 2 columns]

In [ ]:
df_goodreads.head()

bookID                                              title                     authors  average_rating        isbn         isbn13 language_code  num_pages  ratings_count  text_reviews_count publication_date        publisher
0       1  Harry Potter and the Half-Blood Prince (Harry ...  J.K. Rowling/Mary GrandPré            4.57  0439785960  9780439785969           eng        652        2095690               27591        9/16/2006  Scholastic Inc.
1       2  Harry Potter and the Order of the Phoenix (Har...  J.K. Rowling/Mary GrandPré            4.49  0439358078  9780439358071           eng        870        2153167               29221         9/1/2004  Scholastic Inc.
2       4  Harry Potter and the Chamber of Secrets (Harry...                J.K. Rowling            4.42  0439554896  9780439554893           eng        352           6333                 244        11/1/2003       Scholastic
3       5  Harry Potter and the Prisoner of Azkaban (Harr...  J.K. Rowling/Mary GrandPré            4.56  043965548X  9780439655484           eng        435        2339585               36325         5/1/2004  Scholastic Inc.
4       8  Harry Potter Boxed Set  Books 1-5 (Harry Potte...  J.K. Rowling/Mary GrandPré            4.78  0439682584  9780439682589           eng       2690          41428                 164        9/13/2004       Scholastic

###20 How many Lord of the Rings books are available as audiobooks? 

Note: The publisher column has Audio as a subtext in the publisher name.

For example, HarperAudio and Phoenix Audio publish audiobooks. 

In [42]:
df_goodreads[(df_goodreads.publisher.str.contains('udio')) & (df_goodreads.title.str.contains('Lord of the Rings'))]

bookID  ...           publisher
4270   15330  ...  BBC Audiobooks Ltd
4276   15369  ...      BBC Audiobooks

[2 rows x 12 columns]

###Dataset 4. AB_NYC_2019.csv

####7. Which of the following neighborhood is the most expensive in the neighbourhood_group that provides the most number of "Entire home/Apartments" room_type in NYC?

Note: You need to remove null values to get the correct answer.

The dataset can be imported as pandas dataframe using the following link: https://raw.githubusercontent.com/insaid2018/INSAID-Activities/main/Data%20Detective%20Challenge/AB_NYC_2019.csv

In [ ]:
df_neighborhood = pd.read_csv('https://raw.githubusercontent.com/insaid2018/INSAID-Activities/main/Data%20Detective%20Challenge/AB_NYC_2019.csv')

In [ ]:
df_neighborhood.head()

id                                              name  host_id    host_name neighbourhood_group neighbourhood  latitude  longitude        room_type  price  minimum_nights  number_of_reviews last_review  reviews_per_month  calculated_host_listings_count  availability_365
0  2539                Clean & quiet apt home by the park     2787         John            Brooklyn    Kensington  40.64749  -73.97237     Private room    149               1                  9  2018-10-19               0.21                               6               365
1  2595                             Skylit Midtown Castle     2845     Jennifer           Manhattan       Midtown  40.75362  -73.98377  Entire home/apt    225               1                 45  2019-05-21               0.38                               2               355
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632    Elisabeth           Manhattan        Harlem  40.80902  -73.94190     Private room    150               3                  0         NaN                NaN                               1               365
3  3831                   Cozy Entire Floor of Brownstone     4869  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976  Entire home/apt     89               1                270  2019-07-05               4.64                               1               194
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192        Laura           Manhattan   East Harlem  40.79851  -73.94399  Entire home/apt     80              10                  9  2018-11-19               0.10                               1                 0

In [ ]:
df_neighborhood[df_neighborhood['room_type'] == 'Entire home/apt'].groupby(['neighbourhood_group', 'neighbourhood'])['price'].max()

neighbourhood_group  neighbourhood
Bronx                Allerton         450
                     Baychester       101
                     Belmont          299
                     Bronxdale        100
                     Castle Hill      128
                                     ... 
Staten Island        Tottenville      299
                     West Brighton    190
                     Westerleigh      103
                     Willowbrook      249
                     Woodrow          700
Name: price, Length: 216, dtype: int64

####18. How much is the difference between the average pricing of the neighbourhood having the least average pricing of homes and the average pricing of the neighbourhood_group it comes under?

Note: You need to remove null values to get the correct answer.

In [ ]:
df_neighborhood.groupby('neighbourhood')['price'].mean().nsmallest(1)

neighbourhood
Bull's Head    47.333333
Name: price, dtype: float64

In [ ]:
df_neighborhood[df_neighborhood['neighbourhood'] == 'Bull\'s Head'].groupby('neighbourhood_group')['price'].mean()

neighbourhood_group
Staten Island    47.333333
Name: price, dtype: float64

In [ ]:
df_neighborhood

id                                name  host_id host_name neighbourhood_group neighbourhood  latitude  longitude        room_type  price  minimum_nights  number_of_reviews last_review  reviews_per_month  calculated_host_listings_count  availability_365
0  2539  Clean & quiet apt home by the park     2787      John            Brooklyn    Kensington  40.64749  -73.97237     Private room    149               1                  9  2018-10-19               0.21                               6               365
1  2595               Skylit Midtown Castle     2845  Jennifer           Manhattan       Midtown  40.75362  -73.98377  Entire home/apt    225               1                 45  2019-05-21               0.38                               2               355

### Dataset 5. Netflex

####9. Which TV show has the biggest cast and has a valid director name?

The dataset can be imported as pandas dataframe using the following link: https://raw.githubusercontent.com/insaid2018/INSAID-Activities/main/Data%20Detective%20Challenge/netflix_content.csv

In [ ]:
df_netflix = pd.read_csv('https://raw.githubusercontent.com/insaid2018/INSAID-Activities/main/Data%20Detective%20Challenge/netflix_content.csv')

In [ ]:
df_netflix[(df_netflix['type'] =='TV Show') & (df_netflix['director'].notnull())]['num_cast'].nlargest(5).index

Int64Index([322, 2504, 5226, 6961, 2313], dtype='int64')

In [ ]:
df_netflix.head()

show_id     type  title           director                                               cast        country         date_added  release_year rating   duration                                          listed_in                                        description  num_cast           primary_genre
0      s1  TV Show     3%                NaN  João Miguel, Bianca Comparato, Michel Gomes, R...         Brazil    August 14, 2020          2020  TV-MA  4 Seasons  International TV Shows, TV Dramas, TV Sci-Fi &...  In a future where the elite inhabit an island ...        11  International TV Shows
1      s2    Movie   7:19  Jorge Michel Grau  Demián Bichir, Héctor Bonilla, Oscar Serrano, ...         Mexico  December 23, 2016          2016  TV-MA     93 min                       Dramas, International Movies  After a devastating earthquake hits Mexico Cit...         6                  Dramas
2      s3    Movie  23:59       Gilbert Chan  Tedd Chan, Stella Chung, Henley Hii, Lawrence ...      Singapore  December 20, 2018          2011      R     78 min                Horror Movies, International Movies  When an army recruit is found dead, his fellow...         9           Horror Movies
3      s4    Movie      9        Shane Acker  Elijah Wood, John C. Reilly, Jennifer Connelly...  United States  November 16, 2017          2009  PG-13     80 min  Action & Adventure, Independent Movies, Sci-Fi...  In a postapocalyptic world, rag-doll robots hi...         9      Action & Adventure
4      s5    Movie     21     Robert Luketic  Jim Sturgess, Kevin Spacey, Kate Bosworth, Aar...  United States    January 1, 2020          2008  PG-13    123 min                                             Dramas  A brilliant group of students become card-coun...        12                  Dramas

In [ ]:
df_netflix.isnull().sum()

show_id             0
type                0
title               0
director         2389
cast              718
country           507
date_added         10
release_year        0
rating              7
duration            0
listed_in           0
description         0
num_cast            0
primary_genre       0
dtype: int64

####11. Number of TV Shows whose title starts and ends with a H (or h)? 

In [ ]:
#df_netflix['title'][0][0]
df_netflix[ ((df_netflix['title'].str.startswith('H')) |  (df_netflix['title'].str.startswith('h'))) & ((df_netflix['title'].str.endswith('H')) |  (df_netflix['title'].str.endswith('h'))) ]

show_id     type                           title           director                                               cast        country         date_added  release_year rating   duration                                          listed_in                                        description  num_cast             primary_genre
2563   s2564  TV Show                               H                NaN  Jamel Debbouze, Éric Judor, Ramzy Bedia, Cathe...         France  September 1, 2020          2002  TV-MA  4 Seasons                International TV Shows, TV Comedies  At a dysfunctional hospital in Paris, three bu...         6    International TV Shows
2574   s2575    Movie                   Hadi İnşallah  Ali Taner Baltacı  Büşra Pekin, Murat Boz, Cezmi Baskın, Şinasi Y...         Turkey      April 4, 2018          2014  TV-14     95 min    Comedies, International Movies, Romantic Movies  A young woman recovering poorly from a bad rom...        10                  Comedies
2586   s2587    Movie         Halo: The Fall of Reach          Ian Kirby  Jen Taylor, Steve Downes, Michelle Lukes, Brit...  United States      March 7, 2016          2015  TV-14     65 min               Action & Adventure, Sci-Fi & Fantasy  Witness the origins of Master Chief and the Sp...         8        Action & Adventure
2604   s2605    Movie                 Hantu Kak Limah       Mamat Khalid  Awie, Arab Khalid, Sharwani Ns, Tj Isa, Delima...       Malaysia   January 31, 2019          2018  TV-PG    110 min      Comedies, Horror Movies, International Movies  When Kak Limah turns up dead shortly after her...         8                  Comedies
2618   s2619  TV Show                        Happyish                NaN  Steve Coogan, Kathryn Hahn, Sawyer Shipman, Br...  United States       July 1, 2016          2015  TV-MA   1 Season                             TV Comedies, TV Dramas  A middle-aged adman is shaken up by his new bo...         4               TV Comedies
2645   s2646  TV Show                         Hasmukh                NaN  Vir Das, Ranvir Shorey, Manoj Pahwa, Ravi Kish...          India     April 17, 2020          2020  TV-MA   1 Season  Crime TV Shows, International TV Shows, TV Com...  A timid, small-town comedian's long-awaited bi...         8            Crime TV Shows
2676   s2677  TV Show                 Heartbreak High                NaN  Callan Mulvey, Lara Cox, Emma Roche, Ada Nicod...      Australia  November 27, 2020          1999  TV-14  7 Seasons   International TV Shows, TV Dramas, Teen TV Shows  A crash course on the turbulence of being a te...        47    International TV Shows
2756   s2757    Movie                           Hitch       Andy Tennant  Will Smith, Eva Mendes, Kevin James, Amber Val...  United States    January 1, 2020          2005  PG-13    118 min                          Comedies, Romantic Movies  Smooth and sexy Hitch helps clients make a gre...        10                  Comedies
2773   s2774    Movie                    Holiday Rush       Leslie Small  Romany Malco, Sonequa Martin-Green, Darlene Lo...  United States  November 28, 2019          2019  TV-PG     94 min                   Children & Family Movies, Dramas  A widowed radio DJ and his four spoiled kids n...         9  Children & Family Movies
2788   s2789  TV Show  Home: Adventures with Tip & Oh                NaN  Rachel Crow, Mark Whitten, Ana Ortiz, Ron Func...  United States      July 20, 2018          2018  TV-Y7  4 Seasons                              Kids' TV, TV Comedies  A misfit alien named Oh moves in with Tip and ...         6                  Kids' TV
2833   s2834    Movie              House of the Witch        Alex Merkin  Emily Bader, Darren Mann, Michelle Randolph, C...  United States   October 11, 2020          2017  TV-MA     86 min                                      Horror Movies  A group of daring teens finds themselves in a ...         8             Horror Movies
2854   s2855    Movie                      Hridaynath         Amar Gu

In [ ]:
df_netflix[(df_netflix['title'][0]=='H') & (df_netflix['title'][0]=='h')]

KeyError: ignored

#### 15. How many Horror Movies have been released after 2015? 

Note: You need to remove Null values to get the correct answer. 

In [ ]:
df_netflix['primary_genre'].isnull().sum()

0

In [ ]:
df_netflix[(df_netflix['primary_genre'] == 'Horror Movies') & (df_netflix['release_year']>2015)].describe()

release_year    num_cast
count    156.000000  156.000000
mean    2017.858974    7.884615
std        1.277340    2.352655
min     2016.000000    2.000000
25%     2017.000000    6.000000
50%     2018.000000    8.000000
75%     2019.000000   10.000000
max     2020.000000   16.000000

####16. Among the movies released between 1990 and 2000 which of them have been described as a "classic" in their description? 

In [ ]:
df_netflix[(df_netflix.description.str.contains('classic')) & (df_netflix['release_year']>=1990) & (df_netflix['release_year']<=2000)]['description']

1280    Three beauties use their looks, charm and mart...
3076    The Olsen twins put a cute new spin on Mark Tw...
3156    Months after his classic TV sitcom ends, the l...
4370    Hosts Joel and Mike and their robot companions...
Name: description, dtype: object

In [ ]:
df_netflix.head()

show_id     type  title           director                                               cast        country         date_added  release_year rating   duration                                          listed_in                                        description  num_cast           primary_genre
0      s1  TV Show     3%                NaN  João Miguel, Bianca Comparato, Michel Gomes, R...         Brazil    August 14, 2020          2020  TV-MA  4 Seasons  International TV Shows, TV Dramas, TV Sci-Fi &...  In a future where the elite inhabit an island ...        11  International TV Shows
1      s2    Movie   7:19  Jorge Michel Grau  Demián Bichir, Héctor Bonilla, Oscar Serrano, ...         Mexico  December 23, 2016          2016  TV-MA     93 min                       Dramas, International Movies  After a devastating earthquake hits Mexico Cit...         6                  Dramas
2      s3    Movie  23:59       Gilbert Chan  Tedd Chan, Stella Chung, Henley Hii, Lawrence ...      Singapore  December 20, 2018          2011      R     78 min                Horror Movies, International Movies  When an army recruit is found dead, his fellow...         9           Horror Movies
3      s4    Movie      9        Shane Acker  Elijah Wood, John C. Reilly, Jennifer Connelly...  United States  November 16, 2017          2009  PG-13     80 min  Action & Adventure, Independent Movies, Sci-Fi...  In a postapocalyptic world, rag-doll robots hi...         9      Action & Adventure
4      s5    Movie     21     Robert Luketic  Jim Sturgess, Kevin Spacey, Kate Bosworth, Aar...  United States    January 1, 2020          2008  PG-13    123 min                                             Dramas  A brilliant group of students become card-coun...        12                  Dramas

####22. Which of the following movies have been directed by a director who has the most number of Indian movies listed in the dataset?

In [ ]:
df_netflix[df_netflix['director']==df_netflix[df_netflix.country.str.contains('India') & df_netflix.country.notnull()].groupby('director')['director'].count().nlargest(1).index[0]]

show_id   type                              title      director                                               cast country         date_added  release_year rating duration                                          listed_in                                        description  num_cast       primary_genre
1262   s1263  Movie                     Chal Mere Bhai  David Dhawan  Sanjay Dutt, Salman Khan, Karisma Kapoor, Dali...   India  December 31, 2019          2000  TV-14  132 min    Comedies, International Movies, Romantic Movies  When a secretary saves her tycoon boss's life,...        10            Comedies
1286   s1287  Movie                    Chashme Baddoor  David Dhawan  Rishi Kapoor, Ali Zafar, Taapsee Pannu, Siddha...   India       July 5, 2020          2013  TV-14  121 min   Comedies, International Movies, Music & Musicals  When pretty new neighbor Seema falls for their...         9            Comedies
1287   s1288  Movie                    Chashme Buddoor  David Dhawan  Ali Zafar, Siddharth, Divyendu Sharma, Tapsee ...   India        May 1, 2017          2013  TV-PG  121 min   Comedies, International Movies, Music & Musicals  When pretty new neighbor Seema falls for their...        11            Comedies
2643   s2644  Movie               Haseena Maan Jaayegi  David Dhawan  Sanjay Dutt, Govinda, Karisma Kapoor, Pooja Ba...   India      March 1, 2018          1999  TV-PG  151 min               Action & Adventure, Comedies, Dramas  A millionaire opens the door to comic complica...        18  Action & Adventure
3233   s3234  Movie                             Judwaa  David Dhawan  Salman Khan, Karisma Kapoor, Rambha, Kader Kha...   India  December 31, 2019          1997  TV-14  133 min    Comedies, International Movies, Romantic Movies  Separated at birth, a pair of identical twins ...        10            Comedies
3234   s3235  Movie                           Judwaa 2  David Dhawan  Varun Dhawan, Jacqueline Fernandez, Tapsee Pan...   India      March 1, 2018          2017  TV-14  145 min  Action & Adventure, Comedies, International Mo...  Separated at birth, twins Raja and Prem couldn...         8  Action & Adventure
3481   s3482  Movie  Kyo Kii... Main Jhuth Nahin Bolta  David Dhawan  Govinda, Mohnish Bahl, Sharad Kapoor, Satish K...   India       May 16, 2019          2001  TV-PG  150 min   Comedies, International Movies, Sci-Fi & Fantasy  The life and career of a lawyer are thrown int...         8            Comedies
3874   s3875  Movie              Maine Pyaar Kyun Kiya  David Dhawan  Salman Khan, Sushmita Sen, Sohail Khan, Katrin...   India  December 31, 2019          2005  TV-14  137 min   Comedies, International Movies, Music & Musicals  After lying to his girlfriend, orthopedic surg...         8            Comedies
4273   s4274  Movie               Mujhse Shaadi Karogi  David Dhawan  Rajpal Yadav, Amrish Puri, Kader Khan, Shashi ...   India  December 31, 2019          2004  TV-PG  158 min    Comedies, International Movies, Romantic Movies  Samir must overcome a hostile horoscope, a hot...         8            Comedies

###Dataset 6. zomato.csv
####10. What is the Mean Absolute Deviation of the Votes column? 

Note: Make sure you have not dropped Null values. 

The dataset can be imported as pandas dataframe using the following link: https://raw.githubusercontent.com/insaid2018/INSAID-Activities/main/Data%20Detective%20Challenge/zomato.csv

In [ ]:
df_zomato = pd.read_csv('https://raw.githubusercontent.com/insaid2018/INSAID-Activities/main/Data%20Detective%20Challenge/zomato.csv', encoding='latin-1')

In [ ]:
df_zomato['Votes'].mad(axis=0, skipna=False)

193.46493877036863

In [ ]:
df_zomato['Votes'].notnull().sum()

9551

####23. Segregate the third listed cuisine of every restaurant in a new column and calculate which cuisine in that new column has the second-highest frequency of occurrence? 

For example, if a row corresponding to the "Cuisine" column has Bakery, Mexican, North Indian, and Chinese listed in it, then the third listed Cuisine for that particular row would be North Indian.

In [ ]:
df_zomato['new_column'] = df_zomato.Cuisines.str.split(',').str[2]

In [ ]:
totals = df_zomato.groupby('new_column').new_column.count().nlargest(2)
# Note nth is 0 indexed
#second = df_zomato.sort_values('new_column', ascending=False).groupby('new_column').new_column.nth(1)
#ans = pd.DataFrame({'n': totals, 'second': second})
print(totals)

new_column
 Chinese      594
 Fast Food    385
Name: new_column, dtype: int64


Second highest frequency is of Fast Food: 385

### 25 Which of the following arguments can be used to successfully read the Zomato CSV file? 


## I just used encoding='latin-1'

In [ ]:
df_zomato.head()

Restaurant ID         Restaurant Name  Country Code              City                                            Address                                    Locality                                   Locality Verbose   Longitude   Latitude                          Cuisines  Average Cost for two          Currency Has Table booking Has Online delivery Is delivering now Switch to order menu  Price range  Aggregate rating Rating color Rating text  Votes new_column
0        6317637        Le Petit Souffle           162       Makati City  Third Floor, Century City Mall, Kalayaan Avenu...   Century City Mall, Poblacion, Makati City  Century City Mall, Poblacion, Makati City, Mak...  121.027535  14.565443        French, Japanese, Desserts                  1100  Botswana Pula(P)               Yes                  No                No                   No            3               4.8   Dark Green   Excellent    314   Desserts
1        6304287        Izakaya Kikufuji           162       Makati City  Little Tokyo, 2277 Chino Roces Avenue, Legaspi...  Little Tokyo, Legaspi Village, Makati City  Little Tokyo, Legaspi Village, Makati City, Ma...  121.014101  14.553708                          Japanese                  1200  Botswana Pula(P)               Yes                  No                No                   No            3               4.5   Dark Green   Excellent    591        NaN
2        6300002  Heat - Edsa Shangri-La           162  Mandaluyong City  Edsa Shangri-La, 1 Garden Way, Ortigas, Mandal...  Edsa Shangri-La, Ortigas, Mandaluyong City  Edsa Shangri-La, Ortigas, Mandaluyong City, Ma...  121.056831  14.581404  Seafood, Asian, Filipino, Indian                  4000  Botswana Pula(P)               Yes                  No                No                   No            4               4.4        Green   Very Good    270   Filipino
3        6318506                    Ooma           162  Mandaluyong City  Third Floor, Mega Fashion Hall, SM Megamall, O...      SM Megamall, Ortigas, Mandaluyong City  SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.056475  14.585318                   Japanese, Sushi                  1500  Botswana Pula(P)                No                  No                No                   No            4               4.9   Dark Green   Excellent    365        NaN
4        6314302             Sambo Kojin           162  Mandaluyong City  Third Floor, Mega Atrium, SM Megamall, Ortigas...      SM Megamall, Ortigas, Mandaluyong City  SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.057508  14.584450                  Japanese, Korean                  1500  Botswana Pula(P)               Yes                  No                No                   No            4               4.8   Dark Green   Excellent    229        NaN